src/meta/src/manager/background_deleter.rs

```diff
@@ -0,0 +1,176 @@
+// Copyright 2022 Singularity Data
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::sync::Arc;
+
+use itertools::Itertools;
+use risingwave_common::catalog::TableId;
+use risingwave_pb::catalog::{Sink, Source, Table};
+use tokio::sync::oneshot::Sender;
+use tokio::sync::{mpsc, oneshot};
+use tokio::task::JoinHandle;
+
+use super::SourceId;
+use crate::manager::MetaSrvEnv;
+use crate::model::{MetadataModel, MetadataModelResult};
+use crate::storage::MetaStore;
+use crate::stream::{GlobalStreamManagerRef, SourceManagerRef};
+use crate::MetaResult;
+
+pub type CatalogBackgroundDeleterRef = Arc<CatalogBackgroundDeleter>;
+
+macro_rules! create_catalog_deleted {
+    ($name:ident, $cf:ident) => {
+        paste::paste! {
+            #[derive(Debug)]
+            pub struct [<$name Deleted>](pub $name);
+
+            impl MetadataModel for [<$name Deleted>] {
+                type KeyType = <$name as MetadataModel>::KeyType;
+                type ProstType = <$name as MetadataModel>::ProstType;
+
+                fn cf_name() -> String {
+                    $cf.to_string()
+                }
+
+                fn to_protobuf(&self) -> Self::ProstType {
+                    self.0.clone()
+                }
+
+                fn from_protobuf(prost: Self::ProstType) -> Self {
+                    Self($name::from_protobuf(prost))
+                }
+
+                fn key(&self) -> MetadataModelResult<Self::KeyType> {
+                    self.0.key()
+                }
+            }
+        }
+    };
+}
+
+/// Column family name for deleted table catalog.
+const CATALOG_TABLE_DELETED_CF_NAME: &str = "cf/deleted_catalog_table";
+/// Column family name for deleted sink catalog.
+const CATALOG_SINK_DELETED_CF_NAME: &str = "cf/deleted_catalog_sink";
+/// Column family name for deleted source catalog.
+const CATALOG_SOURCE_DELETED_CF_NAME: &str = "cf/deleted_catalog_source";
+
+create_catalog_deleted!(Table, CATALOG_TABLE_DELETED_CF_NAME);
+create_catalog_deleted!(Sink, CATALOG_SINK_DELETED_CF_NAME);
+create_catalog_deleted!(Source, CATALOG_SOURCE_DELETED_CF_NAME);
+
+#[derive(Debug)]
+pub enum CatalogDeletedId {
+    TableId(TableId),
+    SinkId(TableId),
+    SourceId(SourceId),
+}
+
+pub struct CatalogBackgroundDeleter(mpsc::UnboundedSender<Vec<CatalogDeletedId>>);
```

Comment:
Actually the deleter only cares about stream jobs, better to rename it.
```suggestion
pub struct StreamJobBackgroundDeleter(mpsc::UnboundedSender<Vec<CatalogDeletedId>>);
```

Reply:
Better to add some comments explaining what does the deleter do and also what is deleted id.